### PR TITLE
WIP (don't merge!): First stab at compojure-like context feature

### PR DESCRIPTION
**Pull request by zoldar**
This is my first attempt at adding support for a compojure style context. I've spent more time on it than I'd be willing to admit - mainly due to my complete ineptitude when it comes to dealing with macro writing which I hardly do in my own code. The main addition is `context` macro that evaluates routes early with bindings changed accordingly before passing them down to compojure's context. Due to problems in dealing with nested contexts, I had to create a workaround for separately defining subroutes which then get passed into existing context - in the form of `routes` macro, which creates a thunk that has to be explicitly evaluated in the scope of context in which it is embedded. The solution isn't thoroughly tested. However, I wanted to get it out because maybe I'm doing something border-line idiotic while not being aware of it.

This PR is related to issue #17.

### CODE REVIEW

File: src/clojure/clojurewerkz/route_one/compojure.clj [hunk 1]
```diff
@@ -2,41 +2,37 @@
   (:require [compojure.core :as compojure]
             [clojurewerkz.route-one.core :as route-one]))
 
+(def ^{:dynamic true} *route-prefix* nil)
+
+(defmacro handler-with-route [req-handler name path args body]
```
Comment (michaelklishin) [inline src/clojure/clojurewerkz/route_one/compojure.clj]: Please keep argument vectors on a new line.

**Comment by zoldar** [comment]
Sorry, fixed.

**Comment by michaelklishin** [comment]
It would be great if we could avoid the explicit subroute evaluation thing. Have you considered using a stack for active contexts? This is unhygienic (as in macros) but such things are sometimes the only realistic way.

**Comment by zoldar** [comment]
If you could point me at some sort of explanation or an example of using a stack, I would be grateful. Because I have to admit that I have no idea what this can mean in that context (sic!). I thought about attaching metadata which would be then handled accordingly - however this would probably require wrappers for all route related macros/fns - that is, in addition to `routes` and `context` - `defroute`.

**Comment by michaelklishin** [comment]
Context nesting forms a tree, where every new context depends on the previous one. Every new `context` could add some state to a stack which would be accessible to the contexts it nests.

**Comment by michaelklishin** [comment]
Actually, your current solution has a benefit to it: you can reuse contexts fairly trivially. I'd still investigate alternative solutions before we settle on something. We can always come back to this.

**Comment by zoldar** [comment]
Thanks for clearing it up - now I at least have some idea about the stack approrach. Ok, I'll look into it some time later. Thank you for prompt feedback.

**Comment by zoldar** [comment]
I've experimented a bit with unhygienic, stack based approach and I see one important problem - the expansions happen in order determined by the order of declarations - so stack state doesn't reflect the nested structure of contexts. The other one, is the fact that macro expansion of route wrappers seems to happen at a different time than the expansion of context macro. Maybe I'm in the wrong, I'm entirely sure, these are just my observations.

I've put one of (failing) variations of my attempts at this here:

https://github.com/zoldar/route-one/blob/stack-context/src/clojure/clojurewerkz/route_one/compojure.clj

**Comment by zoldar** [comment]
After giving it a bit of a thought, I conculded that approach introduced by me is too hacky to be practical and reliable. It leads to too many exceptions and a lot of places where library user can easily trip up and get confused. I think that it's not worth prusuing further. If you share the sentiment, feel free to close that PR.

**Comment by michaelklishin** [comment]
Fair enough, thank you for your time!

**Pull request closed without merging**